File: apps/backend-java/src/main/java/com/enterpriseagentos/backendjava/domain/common/valueobjects/PersonalityFileName.java
package com.enterpriseagentos.backendjava.domain.common.valueobjects;

import java.util.List;

public final class PersonalityFileName implements Comparable<PersonalityFileName> {
    private final String value;

    public PersonalityFileName(String value) {
        if (value == null || value.isBlank()) {
        throw new IllegalArgumentException("Personality file name must not be empty.");
        }
        if (value.length() > 128) {
        throw new IllegalArgumentException("Personality file name must not exceed 128 characters.");
        }
        value = value.trim();

        this.value = value;
    }

    public String getValue() {
        return value;
}

    public String value() {
        return value;
    }

    public static final List<String> KNOWN_FILE_NAMES = List.of(
        "AGENTS.md",
        "SOUL.md",
        "TOOLS.md",
        "IDENTITY.md",
        "USER.md",
        "BOOTSTRAP.md"
    );

    @Override
    public int compareTo(PersonalityFileName other) {
        return Integer.compare(order(value), order(other.value));
    }

    private static int order(String fileName) {
        for (int index = 0; index < KNOWN_FILE_NAMES.size(); index++) {
            if (KNOWN_FILE_NAMES.get(index).equalsIgnoreCase(fileName)) {
                return index;
            }
        }
        return KNOWN_FILE_NAMES.size() + 1;
    }
}

File: apps/backend/src/Features/Agents/Domain/PersonalityFileName.cs
namespace OffceOs.Domain.Features.Agents;

/// <summary>
/// A validated personality file name from the known set, with composition ordering.
/// </summary>
public readonly record struct PersonalityFileName : IComparable<PersonalityFileName>
{
    private static readonly string[] OrderedNames =
        ["AGENTS.md", "SOUL.md", "TOOLS.md", "IDENTITY.md", "USER.md", "MEMORY.md", "BOOTSTRAP.md"];

    public string Value { get; }

    /// <summary>Composition order index (0-based). Lower = earlier in system prompt.</summary>
    public int CompositionOrder { get; }

    public PersonalityFileName(string value)
    {
        if (string.IsNullOrWhiteSpace(value))
            throw new ArgumentException("Personality file name cannot be empty.", nameof(value));

        value = value.Trim();
        var index = Array.IndexOf(OrderedNames, value);
        if (index < 0)
            throw new ArgumentException($"Unknown personality file: {value}. Must be one of: {string.Join(", ", OrderedNames)}", nameof(value));

        Value = value;
        CompositionOrder = index;
    }

    public static IReadOnlyList<string> KnownFileNames => OrderedNames;

    public int CompareTo(PersonalityFileName other) => CompositionOrder.CompareTo(other.CompositionOrder);

    public override string ToString() => Value;

    public static implicit operator string(PersonalityFileName name) => name.Value;
    public static explicit operator PersonalityFileName(string value) => new(value);
}
